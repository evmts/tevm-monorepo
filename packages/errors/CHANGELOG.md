# @tevm/errors

## 2.0.0-next.105

### Minor Changes

- [#1370](https://github.com/evmts/tevm-monorepo/pull/1370) [`1dcfd69`](https://github.com/evmts/tevm-monorepo/commit/1dcfd6944f77493a00daa0d64590c2b0c0983a0f) Thanks [@roninjin10](https://github.com/roninjin10)! - Renamed tevm/base-client to tevm/node

## 1.1.0-next.100

### Patch Changes

- [#1322](https://github.com/evmts/tevm-monorepo/pull/1322) [`6407be7`](https://github.com/evmts/tevm-monorepo/commit/6407be7736c996aa8939a0ec5ee13c3d3c34f1e5) Thanks [@roninjin10](https://github.com/roninjin10)! - Migrated to vitest for better coverage reporting

## 1.1.0-next.96

### Patch Changes

- [#1301](https://github.com/evmts/tevm-monorepo/pull/1301) [`59268b2`](https://github.com/evmts/tevm-monorepo/commit/59268b2e00423ba8f9ddf6fa89ea0070ae1023a6) Thanks [@roninjin10](https://github.com/roninjin10)! - Added sideEffect: false to package.json for better tree shaking support

## 2.0.0-next.86

### Patch Changes

- [#1240](https://github.com/evmts/tevm-monorepo/pull/1240) [`b1f0504`](https://github.com/evmts/tevm-monorepo/commit/b1f0504a6be9abf08faa029e7b40fb53666293e8) Thanks [@roninjin10](https://github.com/roninjin10)! - Bumped sub dep up

## 2.0.0-next.84

### Patch Changes

- [#1232](https://github.com/evmts/tevm-monorepo/pull/1232) [`a170f0f`](https://github.com/evmts/tevm-monorepo/commit/a170f0f05a624f70cadea95f4fbaf11c00d5cadd) Thanks [@roninjin10](https://github.com/roninjin10)! - Fix bug with error tags not being set properly

- [#1232](https://github.com/evmts/tevm-monorepo/pull/1232) [`a170f0f`](https://github.com/evmts/tevm-monorepo/commit/a170f0f05a624f70cadea95f4fbaf11c00d5cadd) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed issues with error handling and added unit testing

## 2.0.0-next.80

### Patch Changes

- [#1221](https://github.com/evmts/tevm-monorepo/pull/1221) [`b0b63d2`](https://github.com/evmts/tevm-monorepo/commit/b0b63d22076f35d76898ab1094ece9668ceef95d) Thanks [@roninjin10](https://github.com/roninjin10)! - Bump bundler

## 2.0.0-next.79

### Minor Changes

- [#1210](https://github.com/evmts/tevm-monorepo/pull/1210) [`f2d4ac4`](https://github.com/evmts/tevm-monorepo/commit/f2d4ac43dab0c5affe994985851639438cb05911) Thanks [@roninjin10](https://github.com/roninjin10)! - Add compatability for viem code property

### Patch Changes

- [#1211](https://github.com/evmts/tevm-monorepo/pull/1211) [`f51ef40`](https://github.com/evmts/tevm-monorepo/commit/f51ef4007f53b2ca6d4ebff770104d9e9f462ea3) Thanks [@roninjin10](https://github.com/roninjin10)! - Update all packages for new tevm contract changes"

- [#1219](https://github.com/evmts/tevm-monorepo/pull/1219) [`a8070b7`](https://github.com/evmts/tevm-monorepo/commit/a8070b769da6695d5e27569809f8ac86866b081d) Thanks [@roninjin10](https://github.com/roninjin10)! - Making sure every package releases

- [#1213](https://github.com/evmts/tevm-monorepo/pull/1213) [`84a6d9c`](https://github.com/evmts/tevm-monorepo/commit/84a6d9caae5e72246933d72e8721d466b238cf81) Thanks [@roninjin10](https://github.com/roninjin10)! - Upgraded all dependencies except effect to latest

## 1.1.0-next.77

### Patch Changes

- [#1200](https://github.com/evmts/tevm-monorepo/pull/1200) [`398daa0`](https://github.com/evmts/tevm-monorepo/commit/398daa059ed1c4373200da1a114ef07d156b207d) Thanks [@roninjin10](https://github.com/roninjin10)! - Rerelease tevm packages

## 1.1.0-next.75

### Patch Changes

- [#1193](https://github.com/evmts/tevm-monorepo/pull/1193) [`db7bfc7`](https://github.com/evmts/tevm-monorepo/commit/db7bfc7bac341e29e2df20569347eb019e2d37a7) Thanks [@roninjin10](https://github.com/roninjin10)! - Updated server, utils and trie packages. Update readmes. Update jsdoc. Update error handling. Fix corner case bugs found. Add test coverage

## 1.1.0-next.74

### Minor Changes

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Implemented new typesafe error system based on viem errors and the JSON-RPC spec for error codes. All errors come with a typesafe `name` property along with link to docs about the error. They also have a `code` property that maps to ethereum JSON-RPC error codes. All concrete errors are implemented in the `@tevm/errors` package. Each function will then export a union error type of all the errors it and it's sub-functions can throw.

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Renamed action errors like CallError to TevmCallError. They are exported from actions package now instead of errors package

### Patch Changes

- [#1186](https://github.com/evmts/tevm-monorepo/pull/1186) [`7765446`](https://github.com/evmts/tevm-monorepo/commit/7765446beec1391a00f3d3dd8d015d5205e0371a) Thanks [@roninjin10](https://github.com/roninjin10)! - Moved files around to colocate code better. Some packages are disappearing

  - Tevm/Zod is now part of Tevm/actions
  - Tevm/actions-types moved to Tevm/actions
  - Tevm/procedures-types moved to Tevm/procedures

## 1.1.0-next.72

### Minor Changes

- [`4094ead`](https://github.com/evmts/tevm-monorepo/commit/4094eadc105790d4e4046187772a8cdbf28c0ef9) - Fix changesets

## 1.1.0-next.69

### Patch Changes

- [#1163](https://github.com/evmts/tevm-monorepo/pull/1163) [`eb80103`](https://github.com/evmts/tevm-monorepo/commit/eb80103442991a0bbe8342f1e237cb52b1a8cc9e) Thanks [@roninjin10](https://github.com/roninjin10)! - Fix bad publish

## 1.1.0-next.60

### Patch Changes

- [#1127](https://github.com/evmts/tevm-monorepo/pull/1127) [`1314a07`](https://github.com/evmts/tevm-monorepo/commit/1314a0770007dd3aa8a4762ddbec62ac60c1dfb6) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bad release from lack of changeset

## 1.1.0-next.52

### Patch Changes

- [#1088](https://github.com/evmts/tevm-monorepo/pull/1088) [`65e4089`](https://github.com/evmts/tevm-monorepo/commit/65e40891fca12fc7fde5d1e177527cd70b28cb1f) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bug with failing to include changeset for tx package. Bumping every package just to be safe

## 1.1.0-next.47

### Patch Changes

- [#1064](https://github.com/evmts/tevm-monorepo/pull/1064) [`85d9143`](https://github.com/evmts/tevm-monorepo/commit/85d9143262396416a5d94b527f7bd3148ab51069) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed test-utils package being mistakedly private

## 1.1.0-next.45

### Minor Changes

- [#1042](https://github.com/evmts/tevm-monorepo/pull/1042) [`2a00b2f`](https://github.com/evmts/tevm-monorepo/commit/2a00b2fe10171aaa0607aed66c29d8df8c3437c8) Thanks [@roninjin10](https://github.com/roninjin10)! - Added new tevm_mine and anvil_mine support for mining blocks. This includes the JSON-RPC support as well as a new tevm.mine() action to the main tevm api

## 1.0.0-next.40

### Patch Changes

- [#962](https://github.com/evmts/tevm-monorepo/pull/962) [`64c3c6dd09c236c516b117a506380028c3154c4a`](https://github.com/evmts/tevm-monorepo/commit/64c3c6dd09c236c516b117a506380028c3154c4a) Thanks [@roninjin10](https://github.com/roninjin10)! - Added docs for all packages to https://tevm.sh

## 1.0.0-next.28

### Patch Changes

- [#913](https://github.com/evmts/tevm-monorepo/pull/913) [`06268ffeebaec950d1606732c4eb6fd5f9bb77d5`](https://github.com/evmts/tevm-monorepo/commit/06268ffeebaec950d1606732c4eb6fd5f9bb77d5) Thanks [@roninjin10](https://github.com/roninjin10)! - Fixed bug with packages linking to older versions of tevm

## 1.0.0-next.23

### Patch Changes

- [#862](https://github.com/evmts/tevm-monorepo/pull/862) [`f217fa4dc2f730cc109940ef36872229ae9f26d3`](https://github.com/evmts/tevm-monorepo/commit/f217fa4dc2f730cc109940ef36872229ae9f26d3) Thanks [@roninjin10](https://github.com/roninjin10)! - - Renamed MemoryTevm MemoryClient
  - Renamed TevmClient HttpClient
  - Replaced @tevm/actions-types package with @tevm/actions-types, @tevm/client-types, and @tevm/procedures-types packages
  - Moved errors to @tevm/errors
  - Moved bundler packages out of tevm and to @tevm/bundler package
  - Minimized packages exposed in tevm package
  - Fixed bug with missing types exports
