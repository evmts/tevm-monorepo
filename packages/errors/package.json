{
  "name": "@tevm/errors",
  "version": "1.0.0-next.28",
  "private": false,
  "description": "A typesafe library for writing forge scripts in typescript",
  "keywords": [
    "solidity",
    "sol",
    "typescript",
    "web3",
    "blockchain"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/evmts/tevm-monorepo.git",
    "directory": "vm/errors"
  },
  "license": "MIT",
  "contributors": [
    "Will Cory <willcory10@gmail.com>"
  ],
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./types/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "types/index.d.ts",
  "files": [
    "dist",
    "src",
    "!src/**/*.spec.ts",
    "types"
  ],
  "scripts": {
    "all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
    "build": "nx run-many --targets=build:dist,build:types --projects=@tevm/errors",
    "build:dist": "tsup",
    "build:types": "tsup --dts-only && tsc --emitDeclarationOnly --declaration",
    "clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
    "format": "rome format . --write",
    "format:check": "rome format .",
    "generate:docs": "typedoc",
    "lint": "rome check . --apply-unsafe",
    "lint:check": "rome check . --verbose",
    "lint:deps": "bunx depcheck",
    "lint:package": "bunx publint --strict && attw --pack",
    "package:up": "pnpm up --latest",
    "test": "bun test --watch",
    "test:coverage": "bun test --coverage",
    "test:run": "bun test",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@tevm/tsconfig": "workspace:^",
    "@tevm/tsupconfig": "workspace:^"
  },
  "publishConfig": {
    "access": "public"
  }
}
