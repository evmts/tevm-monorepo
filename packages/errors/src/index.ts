export * from './actions/index.js'
export * from './client/index.js'
export * from './ethereumjs/index.js'
export * from './input/index.js'
export * from './procedures/index.js'
export * from './utils/index.js'
export * from './TypedError.js'
export * from './UnexpectedError.js'
