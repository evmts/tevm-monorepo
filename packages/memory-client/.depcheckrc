ignores: [
  "@ethereumjs/evm",
  "@ethereumjs/util",
  "@ethereumjs/block",
  "@ethereumjs/blockchain",
  "@ethereumjs/vm",
  "@tevm/actions-types",
  "@tevm/errors"
]
skip-missing: true
