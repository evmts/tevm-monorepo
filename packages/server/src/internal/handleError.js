/**
 * @param {import('../Client.js').Client} client
 * @param {import('@tevm/errors').BaseError} error
 * @param {import('http').ServerResponse} res
 * @param {{method: string, id?: string | undefined | null | number}} [jsonRpcReq]
 * @returns {void}
 * @throws {never}
 */
export const handleError = (client, error, res, jsonRpcReq = { method: 'unknown' }) => {
	client.transport.tevm.logger.error(error)
	res.writeHead(400, { 'Content-Type': 'application/json' })
	res.end(
		JSON.stringify({
			...(jsonRpcReq.id ? { id: jsonRpcReq.id } : {}),
			method: jsonRpcReq.method,
			jsonrpc: '2.0',
			error: {
				code: error.code,
				message: error.message,
			},
		}),
	)
}
