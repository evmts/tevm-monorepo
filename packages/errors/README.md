<p align="center">
  @tevm/errors
<p>

[![CI](https://github.com/evmts/tevm-monorepo/actions/workflows/e2e.yml/badge.svg)](https://github.com/evmts/tevm-monorepo/actions/workflows/e2e.yml)
[![CI](https://github.com/evmts/tevm-monorepo/actions/workflows/unit.yml/badge.svg)](https://github.com/evmts/tevm-monorepo/actions/workflows/unit.yml)
<a href="https://www.npmjs.com/package/@tevm/errors" target="\_parent">
<img alt="" src="https://img.shields.io/npm/dm/@tevm/errors.svg" />
</a>
<a href="https://bundlephobia.com/package/@tevm/errors@latest" target="\_parent">
<img alt="" src="https://badgen.net/bundlephobia/minzip/@tevm/errors" />
</a>

# @tevm/errors

`@tevm/errors` is the official tevm actions api spec represented as TypeScript types

**This package has no javascript only typescript types.**

The first implementation of this spec is [@tevm/actions](../actions/) which implements it with Ethereumjs.

## Installation

This package is available in the main `tevm` package

```typescript
import {ScriptHandler} from 'tevm/errors'
```

But can also be installed as a standalone package.

```bash
npm install --save-dev @tevm/errors
```

## License ðŸ“„

<a href="./LICENSE"><img src="https://user-images.githubusercontent.com/35039927/231030761-66f5ce58-a4e9-4695-b1fe-255b1bceac92.png" width="200" /></a>
