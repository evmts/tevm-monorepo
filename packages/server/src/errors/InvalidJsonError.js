import { BaseError } from '@tevm/errors'

/**
 * Parameters for constructing an {@link InvalidJsonError}.
 * @typedef {Object} InvalidJsonErrorParameters
 * @property {string} [docsBaseUrl] - Base URL for the documentation.
 * @property {string} [docsPath] - Path to the documentation.
 * @property {string} [docsSlug] - Slug for the documentation.
 * @property {string[]} [metaMessages] - Additional meta messages.
 * @property {import('@tevm/errors').BaseError|Error} [cause] - The cause of the error.
 * @property {string} [details] - Details of the error.
 * @property {object} [meta] - Optional object containing additional information about the error.
 */

/**
 * Represents an error that occurs when parsing JSON fails.
 *
 * This error is typically encountered when there is an issue with the JSON structure, such as a syntax error or malformed JSON.
 *
 * @example
 * try {
 *   const data = parseJsonWithSomeTevmMethod(someString)
 * } catch (error) {
 *   if (error instanceof InvalidJsonError) {
 *     console.error(error.message);
 *     // Handle the invalid JSON error
 *   }
 * }
 *
 * @param {string} message - A human-readable error message.
 * @param {InvalidJsonErrorParameters} [args={}] - Additional parameters for the InvalidJsonError.
 * @property {'InvalidJsonError'} _tag - Same as name, used internally.
 * @property {'InvalidJsonError'} name - The name of the error, used to discriminate errors.
 * @property {string} message - Human-readable error message.
 * @property {object} [meta] - Optional object containing additional information about the error.
 * @property {number} code - Error code, analogous to the code in JSON RPC error.
 * @property {string} docsPath - Path to the documentation for this error.
 * @property {string[]} [metaMessages] - Additional meta messages for more context.
 */
export class InvalidJsonError extends BaseError {
	/**
	 * Constructs an InvalidJsonError.
	 *
	 * @param {string} message - Human-readable error message.
	 * @param {InvalidJsonErrorParameters} [args={}] - Additional parameters for the InvalidJsonError.
	 */
	constructor(message, args = {}) {
		super(
			message,
			{
				...args,
				docsBaseUrl: 'https://tevm.sh',
				docsPath: '/reference/tevm/errors/classes/invalidjsonerror/',
			},
			'InvalidJsonError',
			-32700,
		)
	}
}
