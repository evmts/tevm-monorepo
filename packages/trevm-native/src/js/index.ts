import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};