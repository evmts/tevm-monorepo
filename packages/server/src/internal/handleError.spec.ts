import { InvalidRequestError } from '@tevm/errors'
import { createMemoryClient } from '@tevm/memory-client'
import { describe, expect, it, vi } from 'vitest'
import { handleError } from './handleError.js'

describe('handleError', () => {
	const createMockResponse = () => {
		const res = {
			writeHead: vi.fn(),
			end: vi.fn(),
		}
		return res as any
	}

	it('should handle error with JSON-RPC request id', () => {
		const client = createMemoryClient()
		client.transport.tevm.logger.error = vi.fn() as any
		const error = new InvalidRequestError('Invalid request')
		const res = createMockResponse()
		const jsonRpcReq = { method: 'testMethod', id: 1 }

		handleError(client, error, res, jsonRpcReq)

		expect(client.transport.tevm.logger.error).toHaveBeenCalledWith(error)
		expect(res.writeHead).toHaveBeenCalledWith(400, { 'Content-Type': 'application/json' })
		expect(res.end).toHaveBeenCalledWith(
			JSON.stringify({
				id: jsonRpcReq.id,
				method: jsonRpcReq.method,
				jsonrpc: '2.0',
				error: {
					code: error.code,
					message: error.message,
				},
			}),
		)
		expect(res.end).toMatchSnapshot()
	})

	it('should handle error without JSON-RPC request id', () => {
		const client = createMemoryClient()
		client.transport.tevm.logger.error = vi.fn() as any
		const error = new InvalidRequestError('Invalid request')
		const res = createMockResponse()
		const jsonRpcReq = { method: 'testMethod' }

		handleError(client, error, res, jsonRpcReq)

		expect(client.transport.tevm.logger.error).toHaveBeenCalledWith(error)
		expect(res.writeHead).toHaveBeenCalledWith(400, { 'Content-Type': 'application/json' })
		expect(res.end).toHaveBeenCalledWith(
			JSON.stringify({
				method: jsonRpcReq.method,
				jsonrpc: '2.0',
				error: {
					code: error.code,
					message: error.message,
				},
			}),
		)
		expect(res.end).toMatchSnapshot()
	})

	it('should handle error with default method when JSON-RPC request is not provided', () => {
		const client = createMemoryClient()
		client.transport.tevm.logger.error = vi.fn() as any
		const error = new InvalidRequestError('Invalid request')
		const res = createMockResponse()

		handleError(client, error, res)

		expect(client.transport.tevm.logger.error).toHaveBeenCalledWith(error)
		expect(res.writeHead).toHaveBeenCalledWith(400, { 'Content-Type': 'application/json' })
		expect(res.end).toHaveBeenCalledWith(
			JSON.stringify({
				method: 'unknown',
				jsonrpc: '2.0',
				error: {
					code: error.code,
					message: error.message,
				},
			}),
		)
		expect(res.end).toMatchSnapshot()
	})
})
