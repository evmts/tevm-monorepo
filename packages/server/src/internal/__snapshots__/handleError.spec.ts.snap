// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleError > should handle error with JSON-RPC request id 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "{"id":1,"method":"testMethod","jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request\\n\\nDocs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/\\nVersion: 1.1.0.next-73"}}",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handleError > should handle error with default method when JSON-RPC request is not provided 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "{"method":"unknown","jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request\\n\\nDocs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/\\nVersion: 1.1.0.next-73"}}",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handleError > should handle error without JSON-RPC request id 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "{"method":"testMethod","jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request\\n\\nDocs: https://tevm.sh/reference/tevm/errors/classes/invalidrequesterror/\\nVersion: 1.1.0.next-73"}}",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
